Globals

TimWhiting opened the pull request:
Closes #138

@mannprerak2 
I think I made the needed changes. Tested it locally and it seems to work.

TimWhiting commented:
Okay, this should be ready for review. The only issue right now is the large integration test, which I'm assuming fails because the library maybe has some globals that weren't generated before?

mannprerak2 commented on lib/src/header_parser/sub_parsers/var_parser.dart:
```diff
@@ -0,0 +1,31 @@
+  final type = cursor.type().toCodeGenType();
+  if (type.broadType == BroadType.Unimplemented) {
+    _logger
+        .warning('Global Type not supported $type ${cursor.type().spelling()}');
```
Add these logs instead.
```dart
_logger.fine(
          '---- Removed Global, reason: unsupported type: ${cursor.completeStringRepr()}');
_logger.warning(
      "Skipped global variable '$globalName', type not supported.");
```

TimWhiting replied:
Done

mannprerak2 commented on lib/src/code_generator/global.dart:
```diff
@@ -42,12 +59,9 @@ class Global extends LookUpBinding {
+    final refOrValue = type.broadType == BroadType.Struct ? 'ref' : 'value';
     s.write(
-        "${type.getDartType(w)} get $globalVarName => ($holderVarName ??= ${w.dylibIdentifier}.lookup<${type.getCType(w)}>('$originalName')).value;\n\n");
+        "late final ${type.getDartType(w)} $globalVarName = ${w.dylibIdentifier}.lookup<${type.getCType(w)}>('$originalName').$refOrValue;\n\n");
```
I think global variables are not constants, they can be changed.

If that's the case, we should instead add getters and setters for this.

So for an int global variable, we can do this -
```dart
  late final ffi.Pointer<ffi.Int32> _globalVar =
      _dylib.lookup<ffi.Int32>('globalVar');

  int get globalVar => _globalVar.value;

  set globalVar(int value) => _globalVar.value = value;

```

TimWhiting replied:
Yes, this works for all types except structs, which just don't have a setter, but fields for structs can be set from the ref that is obtained via the getter.

mannprerak2 commented on lib/src/header_parser/translation_unit_parser.dart:
```diff
@@ -58,6 +59,9 @@ int _rootCursorVisitor(clang_types.CXCursor cursor, clang_types.CXCursor parent,
           saveMacroDefinition(cursor);
           break;
+        case clang_types.CXCursorKind.CXCursor_VarDecl:
+          addToBindings(parseVarDefinition(cursor));
```
I think `parseVarDeclaration` will be a more consistent name for this function.

mannprerak2 reviewed (changes requested):
Can you also add a new test in `test/header_parser_tests`? You can pretty much copy the `unnamed_enum_test.dart` file and modify that to add a few tests. (say for a pointer, int, struct, Long Double, pointer to Long double)

mannprerak2 commented:
> The only issue right now is the large integration test, which I'm assuming fails because the library maybe has some globals that weren't generated before?

The `test/debug_generated` folder will contained failed tests, you can compare those with `_expected.*` files  in `large_integration_tests` folder, and subsequently add the changes there if everything seems okay.

TimWhiting commented on lib/src/header_parser/sub_parsers/var_parser.dart:
```diff
@@ -0,0 +1,31 @@
+/// Parses a global variable
+Global? parseVarDefinition(clang_types.CXCursor cursor) {
+  final type = cursor.type().toCodeGenType();
+  if (type.broadType == BroadType.Unimplemented) {
```
Done

TimWhiting commented:
@mannprerak2 
I addressed everything you mentioned, and added a few features that were missing in my previous implementation (the exclude configuration). Added tests, and updated the _expected file in the large integration test.

mannprerak2 commented on lib/src/code_generator/global.dart:
```diff
@@ -34,20 +36,40 @@ class Global extends LookUpBinding {
-        '${w.ffiLibraryPrefix}.Pointer<${type.getCType(w)}> $holderVarName;\n');
-    s.write(
-        "${type.getDartType(w)} get $globalVarName => ($holderVarName ??= ${w.dylibIdentifier}.lookup<${type.getCType(w)}>('$originalName')).value;\n\n");
+        "late final ffi.Pointer<$dartType> $pointerName = ${w.dylibIdentifier}.lookup<$cType>('$originalName');\n\n");
```
`ffi` => `${w.ffiLibraryPrefix}`, this is needed since this prefix will change if some C library has a declaration named `ffi`.

mannprerak2 commented on lib/src/header_parser/sub_parsers/var_parser.dart:
```diff
@@ -0,0 +1,39 @@
+  final type = cursor.type().toCodeGenType();
+  if (type.broadType == BroadType.Unimplemented ||
+      type.getBaseType().isIncompleteStruct ||
+      type.getBaseType().unimplementedReason != null) {
```
You can simply check the BroadType, no need to check if unimplementedReason is null. 
```dart
if (type.getBaseType().broadType == BroadType.Unimplemented ||
    type.getBaseType().isIncompleteStruct){
...
}
```

mannprerak2 commented on lib/src/header_parser/sub_parsers/var_parser.dart:
```diff
@@ -0,0 +1,39 @@
+    return null;
+  }
+  final g = Global(
+    originalName: cursor.spelling(),
```
You also need to add this global to seen globals, If the global is already seen, you can simply return that.
This is needed since the user can include the same header twice, and when that happens, this would generate 2 bindings (say globalVar and globalVar_1).

 You need to define 2 more functions for Globals in lib/src/header_parser/utils.dart file. Similar to this -
```dart

  void addFuncToSeen(String usr, Func func) {
    _functions[usr] = func;
  }

  Func? getSeenFunc(String usr) {
    return _functions[usr];
  }
```

mannprerak2 reviewed (changes requested):
Hi @TimWhiting, Just a few last changes.

mannprerak2 commented on lib/src/header_parser/sub_parsers/var_parser.dart:
```diff
@@ -0,0 +1,39 @@
+        "Skipped global variable '${cursor.spelling()}', type not supported.");
+    return null;
+  }
+  if (!shouldIncludeGlobalVar(cursor.usr(), cursor.spelling())) {
```
Move this up, so that we don't log about globals which the user has excluded.

mannprerak2 reviewed (approved):
lgtm! Thanks @TimWhiting

dcharkes commented on pubspec.yaml:
```diff
@@ -3,23 +3,23 @@
 
 environment:
-  sdk: '>=2.12.0-198.0.dev <3.0.0'
+  sdk: ">=2.12.0-198.0.dev <3.0.0"
```
Maybe undo these formatting changes.

If you want to do formatting changes, please make a separate PR and make the GitHub actions check the formatting.

dcharkes commented on pubspec.yaml:
```diff
@@ -3,23 +3,23 @@
-  cli_util: ^0.2.0    # test->analyzer->cli_util
-  glob: ^1.0.3        # test->analyzer->glob
-  pub_semver: ^1.4.4  # test->analyzer->pub_semver
+  args: ^1.6.0 # Not available
```
Maybe undo these formatting changes.

If you want to do formatting changes, please make a separate PR and make the GitHub actions check the formatting.

dcharkes reviewed (commented):
Cool stuff! Some small comments.

dcharkes commented:
@TimWhiting 

From the CI:

```
00:13 +40: test/header_parser_tests/globals_test.dart: globals_test (setUpAll)                                                                                                                         
WARNING : Skipped global variable 'longDouble', type not supported.
WARNING : Skipped global variable 'pointerToLongDouble', type not supported.
```

Are those on purpose?

TimWhiting commented:
> @TimWhiting
> 
> From the CI:
> 
> ```
> 00:13 +40: test/header_parser_tests/globals_test.dart: globals_test (setUpAll)                                                                                                                         
> WARNING : Skipped global variable 'longDouble', type not supported.
> WARNING : Skipped global variable 'pointerToLongDouble', type not supported.
> ```
> 
> Are those on purpose?

Yes, that is on purpose. I test that the globals whose types dart:ffi does not support are not in the generated bindings and those messages are logged so the user knows those variables will not be present in the generated bindings.
